File: flutter/bicitaxi_conductor/lib/features/rides/models/ride_status.dart
/// Status of a ride throughout its lifecycle.
/// 
/// Canonical status values for Firebase/backend synchronization.
/// All platforms (Flutter & iOS) use the same string values.
enum RideStatus {
  /// Client has requested a ride.
  requested,

  /// System is searching for an available driver.
  searchingDriver,

  /// A driver has been assigned to the ride.
  driverAssigned,

  /// Driver is en route to pickup location.
  driverArriving,

  /// Ride is in progress (passenger on board).
  inProgress,

  /// Ride has been completed successfully.
  completed,

  /// Ride was cancelled.
  cancelled,
}

/// Extension to provide serialization and display names for ride statuses.
extension RideStatusExtension on RideStatus {
  /// Canonical string value for Firebase/backend serialization.
  /// Must match iOS app values exactly.
  String get value {
    switch (this) {
      case RideStatus.requested:
        return 'requested';
      case RideStatus.searchingDriver:
        return 'searchingDriver';
      case RideStatus.driverAssigned:
        return 'driverAssigned';
      case RideStatus.driverArriving:
        return 'driverArriving';
      case RideStatus.inProgress:
        return 'inProgress';
      case RideStatus.completed:
        return 'completed';
      case RideStatus.cancelled:
        return 'cancelled';
    }
  }

  /// Creates a RideStatus from its canonical string value.
  static RideStatus fromValue(String value) {
    switch (value) {
      case 'requested':
        return RideStatus.requested;
      case 'searchingDriver':
        return RideStatus.searchingDriver;
      case 'driverAssigned':
        return RideStatus.driverAssigned;
      case 'driverArriving':
        return RideStatus.driverArriving;
      case 'inProgress':
        return RideStatus.inProgress;
      case 'completed':
        return RideStatus.completed;
      case 'cancelled':
        return RideStatus.cancelled;
      default:
        return RideStatus.requested;
    }
  }
  String get displayName {
    switch (this) {
      case RideStatus.requested:
        return 'Solicitado';
      case RideStatus.searchingDriver:
        return 'Buscando conductor';
      case RideStatus.driverAssigned:
        return 'Conductor asignado';
      case RideStatus.driverArriving:
        return 'Conductor en camino';
      case RideStatus.inProgress:
        return 'En progreso';
      case RideStatus.completed:
        return 'Completado';
      case RideStatus.cancelled:
        return 'Cancelado';
    }
  }

  /// Returns the next status in the workflow (for simulation).
  RideStatus? get nextStatus {
    switch (this) {
      case RideStatus.requested:
        return RideStatus.searchingDriver;
      case RideStatus.searchingDriver:
        return RideStatus.driverAssigned;
      case RideStatus.driverAssigned:
        return RideStatus.driverArriving;
      case RideStatus.driverArriving:
        return RideStatus.inProgress;
      case RideStatus.inProgress:
        return RideStatus.completed;
      case RideStatus.completed:
      case RideStatus.cancelled:
        return null;
    }
  }

  /// Whether the ride is still active (not terminal).
  bool get isActive {
    return this != RideStatus.completed && this != RideStatus.cancelled;
  }
}

File: ios/bicitaxi/bicitaxi/Models/RideStatus.swift
//
//  RideStatus.swift
//  bicitaxi
//
//  Ride lifecycle status enum
//  Canonical values for Firebase/backend synchronization.
//  All platforms (Flutter & iOS) use the same string values.
//

import Foundation

/// Represents the lifecycle status of a ride
/// Raw values match Flutter app for cross-platform Firebase compatibility.
enum RideStatus: String, Codable, CaseIterable {
    case requested = "requested"
    case searchingDriver = "searchingDriver"
    case driverAssigned = "driverAssigned"
    case driverArriving = "driverArriving"
    case inProgress = "inProgress"
    case completed = "completed"
    case cancelled = "cancelled"
    
    /// Human-readable display text
    var displayText: String {
        switch self {
        case .requested: return "Solicitando viaje..."
        case .searchingDriver: return "Buscando conductor..."
        case .driverAssigned: return "¡Conductor asignado!"
        case .driverArriving: return "El conductor está en camino"
        case .inProgress: return "Viaje en progreso"
        case .completed: return "Viaje completado"
        case .cancelled: return "Viaje cancelado"
        }
    }
    
    /// SF Symbol icon for the status
    var iconName: String {
        switch self {
        case .requested: return "clock.fill"
        case .searchingDriver: return "magnifyingglass"
        case .driverAssigned: return "checkmark.circle.fill"
        case .driverArriving: return "bicycle"
        case .inProgress: return "figure.outdoor.cycle"
        case .completed: return "flag.checkered"
        case .cancelled: return "xmark.circle.fill"
        }
    }
    
    /// Whether the ride is still active (not finished)
    var isActive: Bool {
        switch self {
        case .requested, .searchingDriver, .driverAssigned, .driverArriving, .inProgress:
            return true
        case .completed, .cancelled:
            return false
        }
    }
    
    /// Next status in the progression (for simulation)
    var nextStatus: RideStatus? {
        switch self {
        case .requested: return .searchingDriver
        case .searchingDriver: return .driverAssigned
        case .driverAssigned: return .driverArriving
        case .driverArriving: return .inProgress
        case .inProgress: return .completed
        case .completed, .cancelled: return nil
        }
    }
}
